src/yadism/coefficient_functions/__init__.py

```diff
@@ -67,34 +65,23 @@ def light_component(self):
 
         comp = Component(0)
 
-        # the first condition essentially checks nf != 3
-        if nf in masses and masses[nf]:
-            nl = nf - 1
-            comp.extend(
-                fonll.kernels.generate_light(
-                    self.esf,
-                    nl,
-                    self.esf.info.theory["pto_evol"],
-                    self.esf.info.theory["pto"],
-                )
-            )
-            comp.extend(heavy.kernels.generate_missing(self.esf, nl, nl + 1))
+        if self.scheme == "FFN0":
```

Comment:
This does not look correct: light does not depend on any masses and so not on any logs and so is just the same in `FFNS` or `FFN0`

Reply:
The exception is of course missing - but the branching is only needed for those and nowhere else

Reply:
In practice this means `comp.extend(light.kernels.generate(self.esf, nf))` (line 69) should move before the branching and `asy.kernels.generate_light_asy` needs to be renamed to `generate_missing_asy` and rewritten (i.e. it will become much easier

Reply:
The `generate_light_asy` function is just `light.kernels.generate`, `missing`, `damping`, and an additional term at N3LO since $F_d$ doesn't vanish from that order. The  `light.kernels.generate` could indeed be moved out of the asy function, but missing wouldn't be quite the correct name since there is still the N3LO contribution

Actually, this makes me think - we should now also allow for the damping to be applied for FFNS since we may want to that to generate a contribution to $F_d$ in FONLL.

Reply:
I commited something in https://github.com/NNPDF/yadism/pull/195/commits/b828f1222f09781b82bb4d5b8b1a7a48269083c9 (untested) ... I believe this should go into the correct direction ...

Reply:
Thanks! Shouldn't the order as^3 term stay at least somewhere though? And do you agree regarding the damping in FFNS?

Reply:
- concerning damping: asy-s need to be damped for sure, then for the rest I'll write a long comment in a second on the PR
- concerning N3LO: mmm @giacomomagni is the best to comment, but I think in Numerical FONLL there is no need for a special treatment any longer ... @giacomomagni can you remind me what the problem is?

Reply:
> concerning N3LO: mmm @giacomomagni is the best to comment, but I think in Numerical FONLL there is no need for a special treatment any longer ... @giacomomagni can you remind me what the problem is?

I don't know about N3LO specifically, but I agree there should not be any special treatment any longer...

Reply:
I'm not sure I follow entirely the discussion. If I assume $F_d = F^{(nf+1)} -F^{(nf),0} $, 
it is not vanishing already at NNLO (due $C_{q,{NS}}$). 
The only difference at N3LO is that now also $C_g$  and $C_{q,PS}$ should contribute to  $F_d$. 
However we know these coefficients only for the massive $F_{heavy}$ (and only through the approximation of Nicolò), while massive corrections to $C_{i,light}, \ i= {NS,g,PS}$ are unknown.

Reply:
We specifically mean the light coefficient contribution below:
https://github.com/NNPDF/yadism/blob/1bf52faf2eccde3a3049ebd63b94266148c1035d/src/yadism/coefficient_functions/fonll/kernels.py#L50-L58

I don't know where this term originates and if it should be kept (as a contribution to asy?) for numerical fonll.

Reply:
I see, this fix was needed because following the [FONLL paper](https://arxiv.org/pdf/1001.2312.pdf), the cancellation:

https://github.com/NNPDF/yadism/blob/1bf52faf2eccde3a3049ebd63b94266148c1035d/src/yadism/coefficient_functions/fonll/kernels.py#L46-L48

has been implemented. Up to NNLO, it turns out that the light elements can be generated with `nl` instead of `nl+1`.  

This is not true anymore for the N3LO contributions, because the "derivative argument" of the paper (see section 3.2.2 ) do not hold, so one need to properly add the ZM contribution using $F^{(n_l+1)}$. 
This makes a difference because at N3LO the coefficients have an explicit dependence on `nf` which do not factors out anymore (for pure singlet and gluon).

While adding this contribution we skip the `heavy_light` because that piece is already taken care of by: 

https://github.com/NNPDF/yadism/blob/1bf52faf2eccde3a3049ebd63b94266148c1035d/src/yadism/coefficient_functions/fonll/kernels.py#L163

If I understand correctly, in numerical FONLL, the light contribution will be always generated with `nl+1` active flavors, dropping the cancellation done in  FONLL paper;  so this trick should not be needed. 
Hope this can clarify the situation :)

Reply:
Thanks for the clarification, in that case it can indeed be dropped.